Discovery Service RFC (#267)
nuts-foundation/nuts-specification: rfc/rfc022-discovery-service.md

## 3. Service endpoint

The service endpoint is an HTTP endpoint exposed by the server which hosts a list of Verifiable Presentations.
Clients can query the list or submit a new Verifiable Presentation to it.
All presentations MUST conform to the Presentation Definition associated with the discovery service (see Service Definition).

Presentations MUST be encoded in JWT format as string ([JWT]).

### 3.1 Registration

Clients can publish a Verifiable Presentation to the list by sending it in an HTTP POST to the service endpoint.
The HTTP request body MUST be a Verifiable Presentation (in JWT format). The content type of the request MUST be `application/json`.

The server MUST validate the Verifiable Presentation as specified in section 4. If the validation fails, it MUST return a 400 Bad Request response.
If the validation succeeds, the Verifiable Presentation MUST be added to the list and the server MUST return a 201 Created response.

A credential subject (identified by `credentialSubject.id`) MUST NOT appear more than once on the list,

> so a new registration MUST replace the previous one from the same credential subject.

Reviewer: `the list` probably refers to §3.2. This is confusing for the uninitiated

An example posting a Verifiable Presentation in JWT format to the list:

``http request
POST /list HTTP/1.1
Content-Type: application/json

"eyCAFE.etc.etc"
``

The Verifiable Presentation MUST NOT be valid longer than the Verifiable Credentials it contains. 

### 3.2 Reading

Clients MUST read the list by sending an HTTP GET to the server's service endpoint.
The server MUST return a 200 OK response with a JSON object containing:
- `entries` REQUIRED. MUST contain a JSON array with zero or more presentations.

> - `tag` REQUIRED. MUST be a JSON string referring to the last entry. The tag is an opaque value; the client SHOULD NOT derive any meaning from it.

Reviewer: > referring to the last entry.

It refers to the moment in time of the last entry (in the ref impl). A description like: "used by the server to provide delta's" might be better suited

The `tag` query parameter MAY be used by the client to request a delta next time it reads the list.
If no `tag` query parameter is provided, the server MUST return the full list.
Servers MUST only return the latest (valid) presentation per credential subject.

Example:

``http request
GET /list?tag=510 HTTP/1.1
Content-Type: application/json

{
  "entries": [
    "ey1234.etc.etc",
    "ey5678.etc.etc"
  ],
  "tag": "515",
}
``

Clients MUST validate each presentation in the list as specified in section 4.

> If a presentation is valid, the client uses it in its system. If a presentation is not valid, it MUST be rejected.

Reviewer: ```suggestion
If a presentation is not valid, it MUST be rejected.
```

If one or more presentations are not valid, it SHOULD NOT reject the other presentations.

> ### 3.3 List pruning

Reviewer: How does this work with expired (not retracted) VPs? Do clients also have to prune their lists?

- it MUST contain a `retract_jti` JWT claim, containing the `jti` of the presentation to retract.
- it MUST NOT contain any credentials.

If a server receives a retraction that references an unknown presentation it MUST respond with a 400 Bad Request response.
The response MUST be a JSON error response describing the error.

Clients processing a retraction entry MUST remove the presentation indicated by `retract_jti`.

### 3.5 Error responses

If a server encounters an error while processing a request, it MUST respond with a JSON response containing an `error` property, describing the error.

## 4. Presentation processing

To process a presentation, the following validation steps MUST be performed:

- `jti` (JWT ID) of the presentation MUST be a non-empty string.
- the presentation ID  (`jti`) MUST NOT already exist on the list.
- `nbf` (not before) of the presentation MUST have passed.
- `exp` (expiration) of the presentation MUST NOT have passed.
- `exp` MUST be after `nbf`.
- `aud` MUST contain the ID of the service.
- the number of seconds between `nbf` and `exp` MUST NOT exceed `presentation_max_validity` (see Service Definition).

> - all credential issuers MUST be trusted (see section 5). 

Reviewer: implicit through Presentation Definition?

- `endpoint` REQUIRED. String containing the URL where the server hosts the service endpoint.
- `presentation_definition` REQUIRED. JSON object with the Presentation Definition (see [PE]) describing requirements for presentations on the list.
- `presentation_max_validity` REQUIRED. JSON number containing the maximum validity period (number of seconds between `nbf` and `exp`) of a presentation in seconds.

For example:

``json
{
  "id": "uc_university_v1",
  "endpoint": "https://example.com/usecase/university/v1",
  "presentation_max_validity": 259200,
  "presentation_definition": {
    "id": "pd_university",
    "input_descriptors": [
      {
        "id": "pd_university_type",
        "constraints": {
          "fields": [
            {
              "path": ["$.type"],
              "filter": {
                "type": "string",
                "const": "UniversityCredential"
              }
            },
            {
              "path": "$.credentialSubject.name",
              "filter": {
                "type": "string"
              }
            }
          ]
        }
      }
    ]
  }
}
``

## 6. Trust

> Trust of credential issuers (e.g. `did:example:education-accredetor` issuing `EducationalInstitutionCredential`) SHOULD be defined by the presentation definition.

Reviewer: `accreditor`

In this case, there should be 2 constraints: one for the type and one for the issuer:

``json
[
  {
    "path": ["$.type"],
    "filter": {
      "type": "string",
      "const": "EducationalInstitutionCredential"
    }
  },
  {
    "path": "$.issuer",
    "filter": {
      "type": "string",

>       "const": "did:example:education-accredetor"

Reviewer: `accreditor`

    }
  }
]
``

> 

Reviewer: Missing security and privacy considerations

### Using deltas to reduce indexing overhead

When only the full list is available, clients need to validate and re-index (for fulltext search) all presentations.
By having the server return only new presentations instead, clients can only process those new entries.

### Tag implementation

The tag is specified as an opaque value (JSON string). It is up to server to decide how to implement this.

One way to implement this is to use a [Lamport timestamp](https://en.wikipedia.org/wiki/Lamport_timestamp) which is incremented for every presentation received.
That way, timestamps unambiguously reference the last presentation the client received.

A Unix timestamp does not offer this property, as it is possible that multiple presentations are received at the second,
possibly serving clients duplicate presentations.

## References

* [JWT] M. Jones, J. Bradley, N. Sakimura, "JSON Web Token (JWT)", RFC 7519, May 2015, <https://www.rfc-editor.org/info/rfc7519>.
* [VC-DATA-MODEL] M. Sporny, D. Longley, D. Chadwick, "Verifiable Credentials Data Model 1.1", W3C Recommendation, 3 March 2022, <https://www.w3.org/TR/vc-data-model/>.

> * [PE] D. Buchner, B. Zundel, M. Riedel, K.H. Duffy, "Presentation Exchange 2.0.0", 12 September 2023, <https://identity.foundation/presentation-exchange/spec/v2.0.0/>.

Reviewer 2: This is usually referred to as `PEX` since that sounds a lot nicer when pronounced.

* **Client**: system registering presentations on the list and reading it to discover other systems/parties.
* **List**: refers to the list of Verifiable Presentations hosted by the server for a specific service.

Other terms are as defined in the following specifications: "JSON Web Token (JWT)" [JWT], Verifiable Credentials Data Model 1.1 [VC-DATA-MODEL], Verifiable Credentials Presentation Exchange [PEX], Decentralized Identifiers 1.0 [DID].

## 3. Service endpoint

The service endpoint is an HTTP endpoint exposed by the server which hosts a list of Verifiable Presentations.
Clients can query the list or submit a new Verifiable Presentation to it.
All presentations MUST conform to the Presentation Definition associated with the discovery service (see Service Definition).

Presentations MUST be encoded in JWT format as string ([JWT]).

The protocol identifies the server and client roles. A client can be a server and vice versa. The individual operations define what to do in certain cases.

### 3.1 Registration

Clients can publish a Verifiable Presentation to the server by sending it in an HTTP POST to the service endpoint.

> The HTTP request body MUST be a Verifiable Presentation (in JWT format). The content type of the request MUST be `application/json`.

Reviewer: we could drop this

Clients that wish to retain their presence on the list MUST submit a new registration before the current entry's `exp` time has passed.

### 3.4 Retracting presentations

Clients can remove presentations from the list, e.g. when a organization stops supporting the use case.
To remove presentations of a credential subject the client MUST register a Verifiable Presentation as specified in section 3.1,
with the following additional requirements:

- it MUST specify `RetractedVerifiablePresentation` as type, in addition to the `VerifiablePresentation`.
- it MUST contain a `retract_jti` JWT claim, containing the `jti` of the presentation to retract.
- it MUST NOT contain any credentials.

If a server receives a retraction that references an unknown presentation it MUST respond with a 400 Bad Request response.
The response MUST be a JSON error response (§3.5) describing the error.

Clients processing a retraction entry MUST remove the presentation indicated by `retract_jti`.
The expiration of the retraction presentation MUST be equal to the expiration of the presentation to retract.

> ### 3.5 Error responses

Reviewer: why specify the exact error string? There's nothing a user can do. All these errors are caused by a faulty client.

> ### 3.5 Error responses

Reviewer 2: IMO these are 3 different types of errors
- faulty client. action: nothing, or complain to your software provider (make issue for nuts-node)
- misconfigured client. action: users need to figure out who is at fault -> server/client (probably client)
- server dependency failure. action: monitor the issue and take action if it takes too long

The exact error message could be clearer, but I think these categories would help users debug any issues.

> ### 3.5 Error responses

Reviewer 2: probably needs a general `ServerError` too

> ### 3.5 Error responses

Reviewer: in all cases an admin has to do something, not a user.

> ### 3.5 Error responses

Reviewer 2: yes, but the error helps the admin figure out what to do. 

However, these suggestions were the result of o the original issue title in https://github.com/nuts-foundation/nuts-node/issues/3471. The changed scope of that issue makes this irrelevant

## 4. Presentation processing

The abbreviations below are conform definitions by [IANA](https://www.iana.org/assignments/jwt/jwt.xhtml#claims). 
To process a presentation, the following validation steps MUST be performed:

- `jti` of the presentation MUST be a non-empty string.
- `nbf` of the presentation MUST have passed.
- `exp` of the presentation MUST NOT have passed.
- `exp` MUST be after `nbf`.
- `aud` MUST contain the ID of the service.
- the number of seconds between `nbf` and `exp` MUST NOT exceed `presentation_max_validity` (see Service Definition).
- all credentials MUST have the same `credentialSubject.id`.
- `exp` of the presentation MUST NOT be after the expiration date of the credentials.
- the key used to sign the presentation MUST be owned by the credential subject (see 4.1):
  the JWT `kid` header MUST reference an `assertionMethod` key from the subject's DID document.
- The DID of the credential subject MUST be of a DID method defined in the service definition.
- the presentation MAY contain a `DiscoveryRegistrationCredential` credential defined by the Nuts JSON-LD context (https://nuts.nl/credentials/v1).

>   If present, it MUST NOT be included in the next validation steps.

Reviewer 2: what if it is included in the presentation definition (next validation step)?

It makes sense to me that a use-case lists all the endpoints in the PD that are required for the use-case

Reviewer: added an unless.

Reviewer 2: I don't like magic and this also makes it harder to implement. It's unlikely a use-case will only contain the `authServerURL`, so I prefer the credential to be defined explicitly. If needed we can also add a flag to the API to add the `authServerURL` that defaults to true.

- the credentials MUST conform to the Presentation Definition associated with the list (see Service Definition).
  The Verifiable Presentation MUST NOT contain any Verifiable Credentials besides the ones that conform to the Presentation Definition and the `DiscoveryRegistrationCredential`.

If a validation step fails, the presentation MUST be rejected.

> JWT credential and presentation encoding as specified by [VC-DATA-MODEL] MUST be applied.

Reviewer 2: is this saying that credentials also MUST be JWT?

Reviewer: should only be presentation

A clock skew of 5 seconds MAY be applied to `nbf` and `exp` claims.

> ### 4.1 DiscoveryRegistrationCredential

Reviewer 2: add something about missing a proof / sharing proof with the VP

The optional `DiscoveryRegistrationCredential` is used to add additional parameters to the registration.

> It's a _holder_ credential, meaning it has been self issued. Fields under `credentialSubject` (except `id`) can be used by clients depending on the use case.

Reviewer 2: reference `holder credential` concept (Has it been removed from VC data model v2? I can't find it.)

Reviewer: `holder claims`: https://www.w3.org/TR/vc-data-model-2.0/#presentations-including-holder-claims

Reviewer: changed to `self-asserted`

The credential has the following requirements:

> - `type` MUST be `DiscoveryRegistrationCredential`.

Reviewer 2: ```suggestion
- `type` MUST contain `DiscoveryRegistrationCredential`.
```

  "presentation_max_validity": 259200,
  "presentation_definition": {
    "id": "pd_university",
    "input_descriptors": [
      {
        "id": "pd_university_type",
        "constraints": {
          "fields": [
            {
              "path": ["$.type"],
              "filter": {
                "type": "string",
                "const": "UniversityCredential"
              }
            },
            {
              "path": "$.credentialSubject.name",
              "filter": {
                "type": "string"
              }
            }
          ]
        }
      }
    ]
  }
}
``

## 6. Trust

Trust of credential issuers (e.g. `did:example:education-accreditor` issuing `EducationalInstitutionCredential`) SHOULD be defined by the presentation definition.
In this case, there should be 2 constraints: one for the type and one for the issuer:

``json
[
  {
    "path": ["$.type"],
    "filter": {
      "type": "string",
      "const": "EducationalInstitutionCredential"
    }
  },
  {
    "path": "$.issuer",
    "filter": {
      "type": "string",
      "const": "did:example:education-accreditor"
    }
  }
]
``

## 7. Security Considerations

## 7.1 Denial of Service

> Since the server is hosting a public endpoint, it should take the following measures to prevent denial-of-service attacks:

Reviewer 2: Should this apply to all nodes?
All nodes have the endpoint exposed and forward to what they consider the server. (Implementation detail of the nuts-node, so may need to be added to documentation on the node if valid.)

Reviewer: according to the spec, clients don't expose anything.

- rate limiting on the number of requests per client.
- rate limiting on the number of requests per second.
- limit the size of the list.
- limit the size of the presentations that can be uploaded.

Clients should also take measures when the server has been compromised and is hosting malicious content:
- limit the number of presentations to process.
- limit the size of the presentations to process.
- use streaming processing to prevent excessive memory usage.

## 7.2 Trust

> Designers of services should build trust into the Presentation Definition of the server. This can be done by specifying specific issuers of credentials.

Reviewer 2: ```suggestion
Designers of services should write trust into the Presentation Definition of the Service Definition. This can be done by specifying specific issuers of credentials.
```
